### PR TITLE
fix(trends): return interval values AND total volume for clickhouse

### LINKED ISSUE #4858: Math operations not working
Math operations not working (with multiple series?)

Getting `NaN`s and `0`s. Shows correctly on my dashboards.

[Link to table](https://app.posthog.com/insights?insight=TRENDS&events=%5B%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-python%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__web%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A2%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-node%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A3%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-react-native%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A4%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-go%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A5%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-php%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A6%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-android%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A7%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-ios%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A8%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-plugin-server%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%2C%7B%22id%22%3A%22instance%20status%20report%22%2C%22math%22%3A%22dau%22%2C%22name%22%3A%22instance%20status%20report%22%2C%22type%22%3A%22events%22%2C%22order%22%3A9%2C%22properties%22%3A%5B%7B%22key%22%3A%22teams__1__events_count_by_lib__posthog-ruby%22%2C%22type%22%3A%22event%22%2C%22value%22%3A%22is_set%22%2C%22operator%22%3A%22is_set%22%7D%5D%2C%22math_property%22%3Anull%7D%5D&display=ActionsTable&interval=day&date_from=-30d&funnel_window_days=14&properties=%5B%5D#fromItem=167979&fromItemName=Library%20Usage%20Stats&fromDashboard=3102)

Screenshot 2021-06-23 at 12 13 13
Screenshot 2021-06-23 at 12 14 40

### PR DESCRIPTION
**Pull request by hazzadous**
**WARNING: this was meant to be just a fix for the specific case mentioned in #4858 but this change resulted in the updated SQL being in an unexpected format for the formula calculation code, hence only the first commit is actually related to #4858**

Previously we were doing either or, but the frontend is expecting
to have both so that it can do some averaging on the frontend if requested.

It's debatable whether we should be doing this averaging on the backend as
well, and when exactly average on the frontend is meant to be, but I guess
it's supposed to be the mean of the values over the individual intervals.
At least this I think is what the postgres version is doing.

Closes #4858

A note on tests: I have added a regression test for the issue. It is at the REST API boundary level and makes no direct db or python API calls. This should allow us to be much less tied to the implementation and make refactoring lower level code easier. It also has the benefit of making tests partitioned by organisation thus making parallel execution a little easier to implement at some point.

### CODE REVIEW

**Comment by hazzadous** [comment]
@EDsCODE @macobo I've added you both as reviewers but looks like there are some tests failing still. Feel free to leave, although also feel free to comment.

**Comment by hazzadous** [comment]
Just one test failing, test_trends_for_non_existing_action

Haven't figured out why yet.

**Comment by hazzadous** [comment]
@EDsCODE @macobo this is passing now

File: ee/clickhouse/queries/trends/total_volume.py [hunk 1]
```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
...
-            parsed_results = []
-            for _, stats in enumerate(result):
-                parsed_result = parse_response(stats, filter)
-                parsed_results.append(parsed_result)
+            # NOTE: there's some implicit coupling here:
+            #
+            #        1. the ordering of the fields we get back, (date, value)
+            #        2. that the last row is always the `WITH TOTALS` result
+            #
+            #  TODO: remove coupling
```
Comment (macobo) [inline ee/clickhouse/queries/trends/total_volume.py:64]: Who is this TODO for and is this actionable or relevant?

While there is coupling involved here of, so does every bit of querying and query postprocessing. Decoupling them makes things _worse_ more often than not unless the SQL is generic somehow.

File: ee/clickhouse/queries/trends/total_volume.py [hunk 2]
```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
...
+        
+    {aggregate_query}
+
+) aggregate ON aggregate.interval_start = interval_start
+"""
+
+
+def zero_fill(aggregate_query: str) -> str:
+    """
+    WWraps the provided query with a JOIN that will fill in any missing interval
```
Comment (macobo) [inline ee/clickhouse/queries/trends/total_volume.py:126]: Typo

File: ee/clickhouse/queries/trends/total_volume.py [hunk 3]
```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
...
+
+ZERO_FILL_TEMPLATE = """
+SELECT 
+    interval.start AS interval_start,
+
+    -- For where we don't have an aggregate value, we zero fill
+    COALESCE(aggregate.value, 0) AS value
+
+FROM (
+    -- Creates zero values for all date axis ticks for the given date_from, date_to range
```
Comment (macobo) [inline ee/clickhouse/queries/trends/total_volume.py:94]: This is gonna cause issues.

system.query_log automatically strips newlines meaning that this query cannot be copied from there making debugging and future optimizations really annoying.

It can also lead to instances where the SQL parsing gives up due to the SQL query growing too large.

What are these comments trying to achieve? Do they belong within the sql itself?

File: ee/clickhouse/queries/trends/total_volume.py [hunk 4]
```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
...
+-- left join so we end up with values for all intervals, even if we don't have an aggregate
+LEFT JOIN (
+        
+    {aggregate_query}
+
+) aggregate ON aggregate.interval_start = interval_start
+"""
+
+
+def zero_fill(aggregate_query: str) -> str:
```
Comment (macobo) [inline ee/clickhouse/queries/trends/total_volume.py:124]: This introduces a new querying pattern:

1. Query is stored in the same file
2. We use functions instead of `.format` calls to add subqueries

I'd push back on this unless there's a deeper reason here - having too many querying patterns makes code comprehension harder.

File: ee/clickhouse/queries/trends/total_volume.py [hunk 5]
```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
...
+            #
+            #  TODO: remove coupling
+
+            #  The last row relates to the `WITH TOTALS` from the clickhouse query
+            values_without_totals, (_, aggregated_value) = result[:-1], result[-1]
+
+            # Separate dates and values, as desired for the REST API response
+            dates, values = zip(*sorted(values_without_totals))
+
+            parsed_result = parse_response([dates, values], filter)
```
Comment (macobo) [inline ee/clickhouse/queries/trends/total_volume.py:72]: The above 3 lines are much much harder to grok than what was going on previously. Why did this become so complicated? Can we simplify somehow?

File: ee/clickhouse/queries/trends/total_volume.py [hunk 6]
```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
...
 
-            return parsed_results
+            # NOTE: to maintain backwards compatability here from previous behaviour, we exclude the
+            #        `aggregated_value` for some filter.display.
+            # TODO: verify if updating to always including `aggregated_value` would break anything. It
+            #       would be simpler to not have to have this special casing in here.
+            if filter.display in TRENDS_DISPLAY_BY_VALUE:
```
Comment (macobo) [inline ee/clickhouse/queries/trends/total_volume.py:78]: Why isn't this done in parsing the result?

File: posthog/models/entity.py [hunk 7]
```diff
@@ -108,8 +108,8 @@ def get_action(self) -> Action:
         try:
             self._action = Action.objects.get(id=self.id)
             return self._action
-        except:
-            raise ValidationError(f"Action ID {self.id} does not exist!")
+        except Action.DoesNotExist as exc:
+            raise ValidationError(f"Action ID {self.id} does not exist!") from exc
```
Comment (macobo) [inline posthog/models/entity.py:112]: Unrelated change?

File: posthog/api/test/test_trends.py [hunk 8]
```diff
@@ -112,30 +116,106 @@ def test_includes_only_intervals_within_range(client: Client):
...
-            "date_from": request.date_from,
-            "date_to": request.date_to,
-            "interval": request.interval,
-            "insight": request.insight,
-            "breakdown": request.breakdown,
-            "breakdown_type": request.breakdown_type,
-            "display": request.display,
-            "events": json.dumps(request.events),
+            key: value
+            for key, value in dataclasses.asdict(request).items()
```
Comment (macobo) [inline posthog/api/test/test_trends.py:216]: Does this work? events was json dumped before - and it needs to be to actually work.

File: posthog/queries/base.py [hunk 9]
```diff
@@ -40,21 +49,20 @@ def determine_compared_filter(filter) -> Filter:
...
-        entity.update(
-            {
-                "labels": labels,
-                "days": entity["days"],
-                "label": "{} - {}".format(entity["label"], label),
-                "chartLabel": "{} - {}".format(entity["label"], label),
-                "compare": True,
-            }
+    return [
+        dict(
```
Comment (macobo) [inline posthog/queries/base.py:53]: Let's be consistent and use the `{}` syntax over dict to keep the code easy to read.

File: posthog/queries/base.py [hunk 10]
```diff
@@ -40,21 +49,20 @@ def determine_compared_filter(filter) -> Filter:
 
 
 def convert_to_comparison(trend_entity: List[Dict[str, Any]], filter, label: str) -> List[Dict[str, Any]]:
-    for entity in trend_entity:
```
Comment (macobo) [inline posthog/queries/base.py:43]: Is this a pure refactor or did anything in here actually change besides getting rid of the side effect?

File: posthog/queries/base.py [hunk 11]
```diff
@@ -64,7 +72,20 @@ def convert_to_comparison(trend_entity: List[Dict[str, Any]], filter, label: str
 """
 
 
-def handle_compare(filter, func: Callable, team: Team, **kwargs) -> List:
+class InsightResult(TypedDict):
```
Comment (macobo) [inline posthog/queries/base.py:75]: Nice!

File: requirements.in [hunk 12]
```diff
@@ -38,6 +38,7 @@ pickleshare==0.7.5
 posthoganalytics==1.4.2
 protobuf==3.13.0
 psycopg2-binary==2.8.6
+py-expression-eval==0.3.14
```
Comment (macobo) [inline requirements.in:41]: ? Where is this used?

File: posthog/api/test/test_trends.py [hunk 13]
```diff
@@ -112,30 +116,106 @@ def test_includes_only_intervals_within_range(client: Client):
         }
 
 
+@pytest.mark.django_db
+@pytest.mark.ee
+def test_math_is_functional(client: Client):
```
Comment (macobo) [inline posthog/api/test/test_trends.py:121]: What about testing this on postgres?

Also now we have 3 different ways to write tests for queries, each of which has its own strengths which is gonna lead to more issues. What's the plan here?

File: posthog/api/test/test_trends.py [hunk 14]
```diff
@@ -112,30 +116,106 @@ def test_includes_only_intervals_within_range(client: Client):
...
+    signup_ok(client=client, email="user@fake.fake", password="password", organization_name="Hedgehogs United, LLC")
+    project = get_current_project_ok(client=client)
+
+    # "2021-09-19" is a sunday, i.e. beginning of week
+    with freeze_time("2021-09-20T16:00:00"):
+        #  First identify as a member of the cohort
+        distinct_id = "abc"
+        identify(distinct_id=distinct_id, team_id=project["id"])
+
+        capture_event(
```
Comment (macobo) [inline posthog/api/test/test_trends.py:145]: :-1: Not a fan of this pattern at all.

File: posthog/queries/base.py (hunk 11 above)
Comment (macobo) [reply posthog/queries/base.py:75]: This seems unused though.

File: ee/clickhouse/queries/trends/total_volume.py (hunk 1 above)
Comment (hazzadous) [reply ee/clickhouse/queries/trends/total_volume.py:64]: I’ll remove the todo, agree not particularly useful.

On the point of coupling, I think that’s fine as long as the couple is explicit and obvious. Even better if coupling can be picked up by type checker e.g. having an explicit interface defined like InsightResult. Still working on what that could look like though

File: ee/clickhouse/queries/trends/total_volume.py (hunk 3 above)
Comment (hazzadous) [reply ee/clickhouse/queries/trends/total_volume.py:94]: Yes that’s a shame.

comments here are meant to clarify the purpose of sections of the query.

I could break up the query as separate vars and use python comments? I think it will lose some clarity though.

I could also just strip out these comments at runtime?

File: ee/clickhouse/queries/trends/total_volume.py (hunk 4 above)
Comment (hazzadous) [reply ee/clickhouse/queries/trends/total_volume.py:124]: I’m not a fan of just using format here. It’s interface is not obvious (the required args are implicitly specified in the string which makes it tricky to comprehend).

Using a function here I think allows for a bit more clarity, communicates that this is something that can be reused (more so than an arbitrary string var like null_sql for which it is not obvious that I should be coupling my code to), specify the cases that it should be used.

Also happy to move to format

File: posthog/models/entity.py (hunk 7 above)
Comment (hazzadous) [reply posthog/models/entity.py:112]: Yes can revert

File: posthog/queries/base.py (hunk 10 above)
Comment (hazzadous) [reply posthog/queries/base.py:43]: The change here is not mutating the entities. I made the change due to test failure, but I’ll check to see if this actually was the case

File: posthog/queries/base.py (hunk 9 above)
Comment (hazzadous) [reply posthog/queries/base.py:53]: This is specifically creating a new duct from an old with some fields changed. Is there a simple way to do this with {}

File: posthog/queries/base.py (hunk 11 above)
Comment (hazzadous) [reply posthog/queries/base.py:75]: Yeah not used in the file. I did update some of the interfaces here but it cased lots of mypy issues

I’ll move to trends.py(?)

File: requirements.in (hunk 12 above)
Comment (hazzadous) [reply requirements.in:41]: For formula calculations in clickhouse_trends

Comment (macobo) [reply requirements.in:41]: My bad, github automatically collapsed that file.

File: posthog/queries/base.py (hunk 9 above)
Comment (macobo) [reply posthog/queries/base.py:53]:
```{
  **entity
  "foo": "bar"
}```

Comment (hazzadous) [reply posthog/queries/base.py:53]: 👍

File: ee/clickhouse/queries/trends/total_volume.py (hunk 5 above)
Comment (hazzadous) [reply ee/clickhouse/queries/trends/total_volume.py:72]: Right, there are a couple of changes that required updating here

 * I added `WITH TOTALS`
 * this with totals will be the last entry in the results
 * so I changed the query from returning results in the form `[[date1, date2, ...], [value1, value2, ...]]` to `[[date1, value1], [date2, value2], ...]`
 * parse_response is expecting the former format

Some options I could try:

 1. I can check to see if I can get `WITH TOTALS` to work with returning arrays as well
 2. I can change just remove the call to `parse_response` and do something simpler here instead

I'll go with the second unless there's any reason not to

File: ee/clickhouse/queries/trends/total_volume.py (hunk 6 above)
Comment (hazzadous) [reply ee/clickhouse/queries/trends/total_volume.py:78]: Yup I can move it there.

File: posthog/api/test/test_trends.py (hunk 13 above)
Comment (hazzadous) [reply posthog/api/test/test_trends.py:121]: Yup I'll check on postgres.

Yes that is troubling.

For this specific case, where we are testing a specific user scenario I found it useful to go through the process of:

 1. reproducing the issue on app.posthog.com
 2. verifying that it's not a UI issue by poking around in the code
 3. identifying the query that is being sent to the api
 4. writing this test using the API request with my expectation

This seemed like a very simple flow, but I can see that it would also make sense to add tests for the specific module that might be shared by other parts of the system.

File: ee/clickhouse/queries/trends/total_volume.py (hunk 3 above)
Comment (hazzadous) [reply ee/clickhouse/queries/trends/total_volume.py:94]: Look into stripping comments @hazzadous / fstrings?

File: ee/clickhouse/queries/trends/total_volume.py [hunk 15]
```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
...
+                parsed_result["aggregated_value"] = aggregated_value
+
+            return [parsed_result]
 
         return _parse
+
+
+ZERO_FILL_TEMPLATE = """
```
Comment (hazzadous) [inline ee/clickhouse/queries/trends/total_volume.py:86]: Move into sql/... files @hazzadous

File: posthog/api/test/test_trends.py (hunk 14 above)
Comment (hazzadous) [reply posthog/api/test/test_trends.py:145]: Cases as if I was the plugin-server / (event ingest):

 1. Append Event  - create_event()
 2. Create Person + PDI + Append Event - create_person()
 3. Update Person (append new row + group by query)  + Append Event
 4. Create PDI + Append Event
 5. Merge Person (update in postgres, Append new row + group by query) + new PDI

Only need 1 + 2 for testing queries / posthog. Cohorts/actions also required.

@hazzadous to update to use `create_event()` and `create_person()`

File: posthog/api/test/test_trends.py (hunk 13 above)
Comment (hazzadous) [reply posthog/api/test/test_trends.py:121]: ways of writing this type of test:

1. with `django.test.TestCase`
   2. with self.client
   3. using `...Query` objects
2. with pytest function test using DjangoClient

Note: you can use `pytest --log-cli-level=DEBUG`

File: posthog/api/test/test_trends.py (hunk 8 above)
Comment (hazzadous) [reply posthog/api/test/test_trends.py:216]: "events" in the `TrendsRequest` is a string 🤞

**Comment by EDsCODE** [comment]
RE: the comments above relating

I think this testing pattern can make more sense and be more concise than what we have now. I'd agree with @macobo that we should make the person and event creation functions as simple as possible and not try to go through the whole identify logic, but the way that this uses the api and just relies on the clickhouse setting swap to test postgres vs ee is much cleaner than the class overriding we're doing. if we decide to use this pattern and move all the tests we have no into this form, then we won't need class overriding and itll be clear which test is run for ee/postgres/both. the way that ee specific tests run now are non obvious

Also, it does a true api test and covers any logic that we might be doing in the api handlers. Our query classes are actually one level below the api call (which is not really the hugest deal but definitely leaves a weird layer where we have api tests but are "basic" and just dummy tests)

primary concern is that this full api test might be slower than the query class tests?

**Comment by hazzadous** [comment]
I'll close this for now until there's more clarity on what the expectation is here/if this is actually useful.

Re. testing, one key takeaway was that `capture_event` and `identify` isn't a concept that posthog repo should be concerned with, but rather this would sit outside the boundary (maybe as part of plugin-server?). Instead, to correspond with the kafka queue into clickhouse we have create_person and create_event actions (plus correspondingly actions for the redis / postgres implementation).

i.e. removing /e/ from the posthog rest api testing bounardy, and pushing events / persons into the bit at the top in red of:

Testing Strategy diagram drawio (2)
